source/fab/language/__init__.py

```diff
@@ -54,6 +54,30 @@ def products(self) -> List[Path]:
         return []
 
 
+class HashCalculator(Task):
+    def __init__(self, hasher: TextReaderAdler32, database: StateDatabase):
+        self._hasher = hasher
+        self._database = database
+
+    def run(self):
+        for _ in self._hasher.line_by_line():
+            pass  # Make sure we've read the whole file
+        file_info = FileInfoDatabase(self._database)
+        file_info.add_file_info(Path(self._hasher.filename),
+                                self._hasher.hash)
```

Comment:
I may be mistaken in thinking this but I thought by doing it in this way it would preserve the original intended behaviour (i.e. if the hashing reader has already read the file it won't re-read it just to create the hash)... though now that I say that perhaps by putting these on the queue that no longer happens anyway?

Reply:
This is indeed the case - so following some offline discussion I've done as suggested - the `TextReader` decorator that did the hashing before is gone and its logic has been migrated directly into the `HashCalculator` task